config_parser_test.cc

```diff
@@ -1,11 +1,75 @@
+#include <sstream>
+#include <string>
+
 #include "gtest/gtest.h"
 #include "config_parser.h"
 
-TEST(NginxConfigParserTest, SimpleConfig) {
-  NginxConfigParser parser;
-  NginxConfig out_config;
+// Tests that the NginxConfigStatement handles tokens correctly
+TEST(NginxConfigTest, ToString) {
+    NginxConfigStatement statement;
+    statement.tokens_.push_back("foo");
+    statement.tokens_.push_back("bar");
+
+    EXPECT_EQ(statement.ToString(0), "foo bar;\n");
+}
+
+// Creates a test fixture to reduce code duplication
+class NginxStringConfigTest : public ::testing::Test {
+    protected:
+        bool ParseString(const std::string config_string) {
+            std::stringstream config_stream(config_string);
+            return parser_.Parse(&config_stream, &out_config_);
+        }
+        bool ParseFile(const char* file_name) {
+            return parser_.Parse(file_name, &out_config_);
+        }
+        NginxConfigParser parser_;
+        NginxConfig out_config_;
+};
+
+// Tests the output of ParseString
+TEST_F(NginxStringConfigTest, SimpleConfigStream) {
 
-  bool success = parser.Parse("example_config", &out_config);
+    EXPECT_TRUE(ParseString("foo bar;"));
+    EXPECT_EQ(1, out_config_.statements_.size()) << "Config has one statement";
+    EXPECT_EQ("foo", out_config_.statements_.at(0)->tokens_.at(0));
+}
+
+// Tests an invalid configuration with no end token
+TEST_F(NginxStringConfigTest, InvalidConfig) {
+    EXPECT_FALSE(ParseString("foo bar"));
+}
+
+// Tests one nested config
+TEST_F(NginxStringConfigTest, Nested) {
+    EXPECT_TRUE(ParseString("baz { foo bar; }"));
+}
 
-  EXPECT_TRUE(success);
+// Tests 2+ nested configs
+TEST_F(NginxStringConfigTest, DoubleNested) {
+  EXPECT_TRUE(ParseString("qux { baz { foo bar; } }"));
 }
+
+// Tests missing end brace
+TEST_F(NginxStringConfigTest, MissingEndBrace) {
+    EXPECT_FALSE(ParseString("qux { baz { foo bar; } "));
+}
+
+// Tests missing start brace
+TEST_F(NginxStringConfigTest, MissingStartBrace) {
+    EXPECT_FALSE(ParseString("qux { baz { foo bar; } } }"));
+}
+
+// Original test
+TEST_F(NginxStringConfigTest, SimpleConfig) {
+  EXPECT_TRUE(ParseFile("example_config"));
+}
+
+/*
+ *   Tests Full Example Configuration from: 
+ *   https://www.nginx.com/resources/wiki/start/topics/examples/full/
+ */
+TEST_F(NginxStringConfigTest, FullConfig) {
+  EXPECT_TRUE(ParseFile("nginx.conf"));
+}
+
```

Comment:
Consider testing for empty config as per [this post.]( https://piazza.com/class/iwy6115h8ce37q?cid=22) Also there isn't a definitive answer (yet), it is possible functionality we should be concerned about.